src/Reference.php

```diff
@@ -100,59 +100,81 @@ public function getDesiredName(): string
+
+        return $this->getTheirModel($defaults);
+    }
+
     /**
      * Returns destination model that is linked through this reference. Will apply
      * necessary conditions.
      *
      * @param array $defaults Properties
      */
-    public function getModel($defaults = []): Model
+    public function getTheirModel($defaults = []): Model
     {
         // set table_alias
-        if (!isset($defaults['table_alias'])) {
-            if (!$this->table_alias) {
-                $this->table_alias = $this->link;
-                $this->table_alias = preg_replace('/_' . ($this->owner->id_field ?: 'id') . '/', '', $this->table_alias);
-                $this->table_alias = preg_replace('/([a-zA-Z])[a-zA-Z]*[^a-zA-Z]*/', '\1', $this->table_alias);
-                if (isset($this->owner->table_alias)) {
-                    $this->table_alias = $this->owner->table_alias . '_' . $this->table_alias;
-                }
-            }
-            $defaults['table_alias'] = $this->table_alias;
-        }
+        $defaults['table_alias'] = $defaults['table_alias'] ?? $this->getTableAlias();
 
         // if model is Closure, then call it and return model
         if (is_object($this->model) && $this->model instanceof \Closure) {
-            $c = ($this->model)($this->owner, $this, $defaults);
+            $closure = ($this->model)($this->getOurModel(), $this, $defaults);
 
-            return $this->addToPersistence($c, $defaults);
+            return $this->addToPersistence($closure, $defaults);
         }
 
         // if model is set, then return clone of this model
         if (is_object($this->model)) {
-            $c = clone $this->model;
+            $theirModel = clone $this->model;
 
-            return $this->addToPersistence($c, $defaults);
+            return $this->addToPersistence($theirModel, $defaults);
         }
 
         // last effort - try to add model
         if (is_array($this->model)) {
-            $model = [$this->model[0]];
-            $md = $this->model;
-            unset($md[0]);
+            $modelDefaults = $this->model;
+            $theirModel = [$modelDefaults[0]];
 
-            $defaults = array_merge($md, $defaults);
+            unset($modelDefaults[0]);
+
+            $defaults = array_merge($modelDefaults, $defaults);
         } elseif (is_string($this->model)) {
-            $model = [$this->model];
+            $theirModel = [$this->model];
         } else {
-            $model = $this->model;
+            $theirModel = $this->model;
+        }
+
+        if (!$theirModel instanceof Model) {
+            $theirModel = $this->factory($theirModel, $defaults);
         }
 
-        if (!$model instanceof Model) {
-            $model = $this->factory($model, $defaults);
+        return $this->addToPersistence($theirModel, $defaults);
+    }
+
+    protected function getTableAlias(): string
```

Comment:
make it private and name it better - getTableAlias can not return different result than `table_alias`!